### PR TITLE
Fix wmcore_cache file creation in the T0 setup

### LINKED ISSUE #10800: Issue creating /tmp/cmst0/.wmcore_cache file
**Impact of the bug**
T0 Agent not able to create/run any job/workflow using WMCore agent 1.5.2

**Describe the bug**
T0 use the Tier0Feeder component to create the sandbox. When Tier0 started to test a new release based con WMCore 1.5.2, Tier0Feeder component was unable to start properly given the following error:
```
2021-09-13 06:55:27,603:140108589012736:WARNING:Service:The cachefile /tmp/cmst0/.wmcore_cache_xlpsm5ls/cmssdt.cern.ch/-3895757188456107843_GET__SDT_cgi-bin_ReleasesXML_ does not exi
st and the service at https://cmssdt.cern.ch/SDT/cgi-bin/ReleasesXML/ raised a FileNotFoundError(2, 'No such file or directory') when accessed
2021-09-13 06:55:27,673:140108589012736:ERROR:StdBase:About to raise exception <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: [Errno 2] No such file or directory: '/tmp/cmst0/.wmcore_cache_xlpsm5ls/cmssdt.cern.ch/-3895757188456107843_GET__SDT_cgi-bin_ReleasesXML_'
        ClassName : None
        ModuleName : WMCore.WMSpec.WMWorkloadTools
        MethodName : _validateArgFunction
        ClassInstance : None
        FileName : /data/tier0/srv/wmagent/3.0.0/sw.jamadova/slc7_amd64_gcc630/cms/t0/3.0.0/lib/python3.8/site-packages/WMCore/WMSpec/WMWorkloadTools.py
        LineNumber : 105
        ErrorNr : 0
```

**How to reproduce it**
Deploying the T0 (3.0.0) agent and start all the components of the agent or only Tier0Feeder.

**Expected behavior**
The /tmp/cmst0/.wmcore_cache file must be correctly created and finally Tier0Feeder will be able to create the sandbox

**Additional context and error message**
Issue is **only** affecting when T0 use WMCore 1.5.2.

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #10800 
Superseeds https://github.com/dmwm/WMCore/pull/10801
Superseeds https://github.com/dmwm/WMCore/pull/10888

#### Status
ready

#### Description
This problem hasn't been completely understood, but it comes from the StdSpecs validation using the TagCollector class. The object likely goes out of scope, which kicks in the garbage collection of the file and the deletion of its cache directory through the `TempDirectory` class.

The proposed way forward - without any behavior change for the base class - here is to define a cache directory *_CACHE_DIR  env var (or WMCORE_CACHE_DIR) specifying which directory to use for all the cache files. Further details can be found at the docstring being provided in the PR.

Last but not least, it provides some back pylint fixes.

#### Is it backward compatible (if not, which system it affects?)
yes

#### Related PRs
None

#### External dependencies / deployment changes
None

### CODE REVIEW

**Comment by jhonatanamado** [comment]
Hi @amaltaro ,Alan I left as reviews the output of this patch. Just two comments:

1. I had to change the `print` with `self['logger'].info`
2. The traceback is the following:
```
2021-12-14 17:59:49,410:139686246188800:WARNING:Service:The cachefile /tmp/cmst0/.wmcore_cache__es3cfx5/cmssdt.cern.ch/7232191333999380798_GET__SDT_cgi-bin_ReleasesXML_ does not exist and the service at 
https://cmssdt.cern.ch/SDT/cgi-bin/ReleasesXML/ raised a FileNotFoundError(2, 'No such file or directory') when accessed
2021-12-14 17:59:49,412:139686246188800:ERROR:StdBase:About to raise exception <@========== WMException Start ==========@>
Exception Class: WMSpecFactoryException
Message: [Errno 2] No such file or directory: '/tmp/cmst0/.wmcore_cache__es3cfx5/cmssdt.cern.ch/7232191333999380798_GET__SDT_cgi-bin_ReleasesXML_'
        ClassName : None
        ModuleName : WMCore.WMSpec.WMWorkloadTools
        MethodName : _validateArgFunction
        ClassInstance : None
        FileName : /data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/WMSpec/WMWorkloadTools.py
        LineNumber : 141
        ErrorNr : 0

Traceback: 
  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/WMSpec/WMWorkloadTools.py", line 132, in _validateArgFunction
    if not valFunction(value):

  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/WMSpec/StdSpecs/StdBase.py", line 1016, in <lambda>
    "CMSSWVersion": {"validate": lambda x: x in releases(),

  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/ReqMgr/Tools/cms.py", line 219, in releases
    return TC.releases(arch)

  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/Services/TagCollector/TagCollector.py", line 89, in releases
    for row in self.data():

  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/Services/TagCollector/TagCollector.py", line 81, in data
    data = self._getResult()

  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/Services/TagCollector/TagCollector.py", line 66, in _getResult
    f = self.refreshCache(cFile, callname, args, encoder=encoder, decoder=decodeBytesToUnicode,

  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/Services/Service.py", line 206, in refreshCache
    self.getData(cachefile, url, inputdata, incoming_headers, encoder, decoder, verb, contentType)

  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/Services/Service.py", line 335, in getData
    raise he

  File "/data/tier0/srv/wmagent/3.0.2/sw/slc7_amd64_gcc630/cms/t0/3.0.2/lib/python3.8/site-packages/WMCore/Services/Service.py", line 311, in getData
    with open(cachefile, 'w') as f:

<@---------- WMException End ----------@>
```
Let me know if you need more information

**Comment by amaltaro** [comment]
@jhonatanamado thanks for testing this. Can you please place the full ComponentLog somewhere in your webarea or afs?
My first commit was using the logger object, but somehow logs were not getting recorded in ReqMgr2, hence I moved them to print. Thanks for figuring this out!

**Comment by jhonatanamado** [comment]
Hi @amaltaro , here the full log `/afs/cern.ch/user/c/cmst0/public/ErrorLogs/WMCore/cache_file`

**Comment by amaltaro** [comment]
Thanks for testing this, @jhonatanamado. We still do not have enough information though.
I have done some updates to this patch, could you please run a new replay from scratch? This patch needs to be applied before any component is started (I also updated the print statements to use the logger object).

**Comment by germanfgv** [comment]
@amaltaro You can find another component log here:
```
/afs/cern.ch/user/c/cmst0/public/ErrorLogs/WMCore/cache_file/ID211215130847
```

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
YESSSSSS!!

**Comment by amaltaro** [comment]
@germanfgv @jhonatanamado could you please trigger a new replay with this patch in? I am no longer able to access the previous log.
As usual, we do not need any load on the replay node, as soon as it hits this exception, you share the logs and we do not need anything else from the replay. Thank you very much!

**Comment by jhonatanamado** [comment]
Hello @amaltaro, Alan, as requested here Logs, one with the patch and the other without this patch.
`/afs/cern.ch/user/j/jamadova/public/WMCore/CacheIssue`

While I was starting the agent the Tier0Feeder component shows the following:
```
started with pid 2127048
Starting : Tier0Feeder
AMRR creating directory: /tmp/cmst0/.wmcore_cache_u2_rbyvd/cmssdt.cern.ch
AMRR SUCCEEDED creating directory: /tmp/cmst0/.wmcore_cache_u2_rbyvd/cmssdt.cern.ch
AMRR cachepath in self: True and req_cache_path: /tmp/cmst0/.wmcore_cache_u2_rbyvd/cmssdt.cern.ch/.cache
Tier0Feeder.__init__
Starting Tier0Feeder as a daemon 
Log will be in /data/tier0/srv/wmagent/3.0.3/install/tier0/Tier0Feeder 
Waiting 1 seconds, to ensure daemon file is created
```
Let me know if you need more info

**Comment by amaltaro** [comment]
@jhonatanamado @germanfgv Jhonatan, apologies for the belated follow up.

I had a look at the logs and I think there is a simple way to ensure that the TagCollector directory exists before we try to write out that cache file in the file system. My last commit here has a 2-lines change for the TagCollector which will hopefully fix this problem. Could you please trigger another replay and share the outcome/logs with me? Thanks

**Comment by amaltaro** [comment]
@jhonatanamado wait, unit tests are pretty bad and it looks like I have to fix something else first. I will let you know as soon as it's ready to be tested.

**Comment by amaltaro** [comment]
@jhonatanamado @germanfgv feel free to test it at your convenience.

**Comment by jhonatanamado** [comment]
Hello @amaltaro, Alan, here Logs,
`/afs/cern.ch/user/j/jamadova/public/WMCore/CacheIssue/20-03-22`

While I was starting the agent the Tier0Feeder component shows the following:
```
Starting : Tier0Feeder
AMRR creating directory: /tmp/cmst0/.wmcore_cache_ptagboz_/cmssdt.cern.ch
AMRR SUCCEEDED creating directory: /tmp/cmst0/.wmcore_cache_ptagboz_/cmssdt.cern.ch
AMRR cachepath in self: True and req_cache_path: /tmp/cmst0/.wmcore_cache_ptagboz_/cmssdt.cern.ch/.cache
AMRR creating directory: /tmp/cmst0/.wmcore_cache_ptagboz_/cmssdt.cern.ch/cmssdt.cern.ch
AMRR SUCCEEDED creating directory: /tmp/cmst0/.wmcore_cache_ptagboz_/cmssdt.cern.ch/cmssdt.cern.ch
Tier0Feeder.__init__
Starting Tier0Feeder as a daemon 
Log will be in /data/tier0/srv/wmagent/3.0.3/install/tier0/Tier0Feeder 
Waiting 1 seconds, to ensure daemon file is created
started with pid 3719495
```
Let me know if you need more info

**Comment by amaltaro** [comment]
@jhonatanamado Jhonatan, can you please apply the suggestion I made in this comment:
https://github.com/dmwm/WMCore/issues/10800#issuecomment-995035488

and trigger another replay?

In short, add something like:
```
export WMCORE_CACHE_DIR=/data/cache
```
to the agent env.sh. Note that this needs to be done before you actually source this `env.sh` script and start the components.

**Comment by jhonatanamado** [comment]
Hi @amaltaro , Indeed is working with `export WMCORE_CACHE_DIR=/tmp/cmst0` on our `env.sh` script.

**Comment by amaltaro** [comment]
Cool! I thought it could be related to something with the `/tmp` directory, but it worked fine when you explicitly define it, which makes me think the problem lies with the temporary directory created in the Requests.py module, these two lines:
```
            idir = tempfile.mkdtemp(prefix='.wmcore_cache_')
            self['deleteCacheOnExit'] = TempDirectory(idir)
```

Let me try one last change now, removing the `TempDirectory` wrapper class (in the last commit: "No longer delete cache directory on exit")

@jhonatanamado can you please remove the environment variable, apply this patch and trigger another replay?

**Comment by jhonatanamado** [comment]
hello @amaltaro, Alan, after removing the environment variable the replay is running and seems to be working too.

**Comment by amaltaro** [comment]
@jhonatanamado confirmed over slack that the current proposed code actually fixes the problem they have been having in the T0 WMAgent over the last many months. Let me just clean those simple pylint messages though.

File: src/python/WMCore/Services/Requests.py [hunk 1]
```diff
@@ -367,7 +369,9 @@ def cacheTopPath(self, given_path, service_name):
                 break
         else:
             idir = tempfile.mkdtemp(prefix='.wmcore_cache_')
-            self['deleteCacheOnExit'] = TempDirectory(idir)
+            # Alan Malta in 29 Mar 2022: this seems to prematurely remove the cache
+            # directory. For details, see: https://github.com/dmwm/WMCore/pull/10915
+            # self['deleteCacheOnExit'] = TempDirectory(idir)
```
Comment (vkuznet) [inline src/python/WMCore/Services/Requests.py:374]: it is unclear to me how this change will affect the rest of the running services/app. If you do not delete temp directory who will take care of it? And, how? I rather prefer to see full description how this should be done. And, so far this PR does not address this. You may well been fixed the current issue but not handling temp area may cause another one, when eventually you'll fill out the disk with non empty temp directories.

**Review by vkuznet — COMMENTED** [review]
Even though changes are trivial I do not understand the logic of the changes with respect of running services/apps. I suggest that you'll clarify how removal of TempDirectory will affect long running services.

File: src/python/WMCore/Services/Requests.py (hunk 1 above)
Comment (amaltaro) [reply src/python/WMCore/Services/Requests.py:374]: The `/tmp` directory has an intrinsic temporary nature, reason why we actually do not use /tmp for our services deployed in CMSWEB. In other words, at least in Unix systems, we cannot guarantee that a temporary file/directory will remain available for us.

For further details, please have a look at the first description of this PR. It could be that I updated it while you were looking at it.

**Comment by amaltaro** [comment]
> Even though changes are trivial I do not understand the logic of the changes with respect of running services/apps. I suggest that you'll clarify how removal of TempDirectory will affect long running services.

In addition to my reply along the code, here it goes. 
Besides T0 WMAgent, these changes have not impact in any of the WMCore central services because all of those define a *_CACHE_DIR, which does not use `TempDirectory`.

**Review by todor-ivanov — APPROVED** [review]
Thanks @amaltaro, me similarly to Valentin have the same concern about not cleaning temporary caches and directories. But since you mention this in the PR comments:
> This is the operation mode for all our central services, and it has never given us any issues

I'd trust your experience here.

File: src/python/WMCore/Services/Requests.py (hunk 1 above)
Comment (vkuznet) [reply src/python/WMCore/Services/Requests.py:374]: I'm not sure it answer my question. The cache is not related to `/tmp` in any way, but you commented out `dleteCacheonExit` key. Based on its name I would conclude that after some operations the cache is not cleaned. As with any disk space it will keep growing until someone will do a clean-up. Therefore my question is not about `TempDirectory` or `/tmp` but rather how the cache is cleaned and where. A simple statement that we never experience an issue it does not mean that eventually disk will fill out and you will end-up with the filled disk issue. What I asked is to clarify how cache clean-up is handled.

**Comment by amaltaro** [comment]
Valentin, a simple answer for your inline question is:
* if CACHE_DIR is defined: there is no procedure for cleaning cache, neither automated nor manual. This PR does not change any behavior on that front.
* else - random cache, meaning `/tmp` in Linux - then the cache is meant to clean after itself once the object using it is garbage collected. This PR **does** change that behavior such that it no longer cleans anything.

As I said before, central services define CACHE_DIR, meaning no clean up was ever performed. Cache in the central services follow something like `/data/srv/state/reqmgr2` (or replace reqmgr2 by the service name using the cache), and I do not see more than a few tens of cache files in there, in my VM). This change is also NOT going to affect any of the central services, as I already mentioned a couple of times above.

The only affected system will be T0 WMAgent, which already has ~50k krb5 tmp files in their `/tmp` area anyways.

In short, I think it's a valid concern, but history and experience doesn't raise any concerns on that front, so I'd rather move forward with this.

**Comment by vkuznet** [comment]
Alan, thanks for providing this info, indeed this clarifies things, and my suggestion was exactly to put such description into this PR to avoid possible confusion in a future. I still suggest to update file in question and add this description about the CACHE either in file's docstring or directly in place where code relies on CACHE. I will leave it up to you though. From code changes point of view everything is clean on my side.

**Comment by amaltaro** [comment]
This is a good point and I agree that documenting this would be useful!

Giving it a second though, we have a functional alternative for this problem experienced by the T0, which is simply defining a CACHE_DIR at their environment as well. With that in place, none of these changes are required and we can leave the TempDirectory delete-on-exit feature around, even if it's not fully functional. Any thoughts?

**Comment by vkuznet** [comment]
I think the proper way to do that is to perform the following set of actions:
- document code about logic of CACHE and usage of env variables
- check in a code the CACHE env if it is not set put logger warning message that code will use TempDirectory class which and provide details what it means
- if delete-on-exitis present you should also provide appropriate logger warning message about its actions, e.g. provide location, set of files which will be deleted, etc. This should include message before the deletion and provide details after the deletion.

This, at least will provide enough details to debug the issue.

**Comment by amaltaro** [comment]
@jhonatanamado Hi Jhonatan, as you can see from above discussions, we decided that the best would be to keep the cache cleanup in place and instead use the environment variable that we also use in central services.

> Hi @amaltaro , Indeed is working with export WMCORE_CACHE_DIR=/tmp/cmst0 on our env.sh script.

Could you please run one last replay, by defining the `WMCORE_CACHE_DIR` env var and applying this patch? This patch won't do anything actually, other than document some of these findings.

**Comment by amaltaro** [comment]
@vkuznet some of what you asked above is not possible because the deletion actually happens when the object is "destroyed", meaning that we likely no longer have access to the logger object. However, I did document it at the file docstring level. I also reverted the removal of the TempDirectory. You might want to have another look into it.

**Review by vkuznet — APPROVED** [review]

**Comment by jhonatanamado** [comment]
> @jhonatanamado Hi Jhonatan, as you can see from above discussions, we decided that the best would be to keep the cache cleanup in place and instead use the environment variable that we also use in central services.
> 
> > Hi @amaltaro , Indeed is working with export WMCORE_CACHE_DIR=/tmp/cmst0 on our env.sh script.
> 
> Could you please run one last replay, by defining the `WMCORE_CACHE_DIR` env var and applying this patch? This patch won't do anything actually, other than document some of these findings.

Hello @amaltaro , Alan as reported before, using `WMCORE_CACHE_DIR=/tmp/cmst0` in our env works as expected.

**Comment by amaltaro** [comment]
Thanks, @jhonatanamado . Then please make sure this variable is defined in the environment script (or in the manage script) such that T0 WMAgents (also replay ones) always have it defined. FYI @germanfgv 

Let me squash these changes and get it merged (thanks for the review as well!)
